### ISSUE TITLE
The worker is not freed when an unresolved future loses its reference.

### ISSUE THREAD

**Issue by randy3k** [HenrikBengtsson/future.callr#20]
```
future::plan(future.callr::callr, workers = 10)

future::nbrOfFreeWorkers()
#> [1] 10

x <- future.callr::callr(1)
x <- NULL

future::nbrOfFreeWorkers()
#> [1] 9
```

**Comment by randy3k** [comment]
One workaround that I found is to clear the `db` object manually.
```r
assign("db", list(), envir = environment(environment(future.callr:::nbrOfFreeWorkers.callr)$FutureRegistry))
```

I guess you could make use of [weak references](https://rlang.r-lib.org/reference/new_weakref.html) to keep track of the futures in the registry.

**Comment by HenrikBengtsson** [comment]
Thanks for reporting.  Yes, "this is expected".  Modulo issue #21, at least it should always be possible to create one more future in this situation, because when `nbrOfFreeWorkers() == 0`, then (and only then), an attempt to resolve and collect (="free up") one of the active futures.

Before anything else, I'm curious, how do you end up in this situation?  

> 1.  If the future is finished, we should remove the future from the registry

Yes, one could hook into R's garbage collector by registering a finalizer using one dummy environment per future.  When finalized, we could free up the internal registry.  FWIW, this is what happens when we lose references to R connections; they're cleaned out when running the GC.

> 2. If the future is still running, there should be a mechanism to stop the R subprocess

This is possible but less straightforward because it has to fit in with the Future API, which yet doesn't have a concept of _terminating_ futures.  It also introduces a side effect, i.e. terminating instead of letting a future expression complete.  So, these type of behaviors must work the same regardless of future backend.  It's no [the roadmap](https://www.futureverse.org/roadmap.html) to figure this out.

**Comment by randy3k** [comment]
I noticed that no workers were available after I have interrupted a call to `future.apply::future_lapply`.

```
r$> future::plan(future.callr::callr, workers = 5)

r$> future.apply::future_lapply(1:5, function(.) Sys.sleep(1000))
^C

r$> future::nbrOfFreeWorkers()
[1] 0
```

**Comment by randy3k** [comment]
But I just realized that it is a more general problem as it also affects `multisession`
```
r$> future::plan(future::multisession, workers = 5)

r$> future.apply::future_lapply(1:5, function(.) Sys.sleep(1000))
^C

r$> future::nbrOfFreeWorkers()
[1] 0
```

**Comment by randy3k** [comment]
> Yes, one could hook into R's garbage collector by registering a finalizer using one dummy environment per future. When finalized, we could free up the internal registry. FWIW, this is what happens when we lose references to R connections; they're cleaned out when running the GC.

If the registry is keeping a reference of the futures, the futures will never be gc'ed. That's why I mentioned weak reference a bit earlier.

**Comment by HenrikBengtsson** [comment]
# Workaround

I don't I ever explained how to get out:

```r
library(future.apply)
plan(multisession, workers = 2)
y <- future_lapply(1:2, function(.) Sys.sleep(1000)) ## Ctrl-C
#> ^C
nbrOfFreeWorkers()
#> [1] 0
```

where there are no more available workers.  To get out of this state, we currently have to force a new set of workers;

```r
plan(sequential)
plan(multisession, workers = 2)
nbrOfFreeWorkers()
#> [1] 2
```

# Moving forward

Having said this, I think there's more that can be done by the **future** framework here.  For example, after the original futures finished, which takes 1000 seconds, we should be able to create new futures, but currently we can't.

One approach could be to check if `nbrOfFreeWorkers() == 0L`, then an attempt to garbage collect internal futures should be made before creating the first future.

**Comment by randy3k** [comment]
The workaround only works with multisession, but not callr workers.

**Comment by HenrikBengtsson** [comment]
> The workaround only works with multisession, but not callr workers.

Correct; I forgot about this. There's a similar problem when creating 'cluster' future with a BYO 'cluster';

```sh
> cl <- parallelly::makeClusterPSOCK(2L)
> plan(cluster, workers = cl)
> for (kk in 1:2) future(Sys.sleep(1000))
> nbrOfFreeWorkers()
[1] 0
> plan(sequential)
> nbrOfFreeWorkers()
[1] 1
> plan(cluster, workers = cl)
# stalls until one of the `cl` workers is free
```

I understand why this happens in both cases.  It is actually only for `plan(multisession)`, and `plan(cluster, workers = nworkers)` and some other cases where the `plan(sequential)` workaround works.

**Comment by HenrikBengtsson** [comment]
Some updates on this. The latter example of restarting future backends with unresolved futures has been addressed in **future** 1.40.0 that is now on CRAN;

```r
library(future)
cl <- parallelly::makeClusterPSOCK(2L)
plan(cluster, workers = cl)
for (kk in 1:2) future(Sys.sleep(1000))
nbrOfFreeWorkers()
#> [1] 0
plan(sequential)
plan(cluster, workers = cl)
nbrOfFreeWorkers()
#> [1] 2
```

In the _next_ release of **future.callr**, this will also work for `callr` futures, e.g.

```r
library(future)
plan(future.callr::callr, workers = 2)
for (kk in 1:2) future(Sys.sleep(1000))
nbrOfFreeWorkers()
#> [1] 0
plan(sequential)
plan(future.callr::callr, workers = 2)
nbrOfFreeWorkers()
#> [1] 2
```

**Comment by HenrikBengtsson** [comment]
More updates. The fact that the non-references future in the original example occupies one worker as reflected by `nbrOfFreeWorkers()` is _not_ major problem, because the worker is indeed available if needed. Here's an illustration of that:
```r
library(future)
plan(multisession, workers = 2)
stopifnot(nbrOfFreeWorkers() == 2)

f <- future(42)
stopifnot(nbrOfFreeWorkers() == 1)

## Drop reference to the future
f <- NULL

## The future still occupies a worker;
stopifnot(nbrOfFreeWorkers() == 1)

## Futures are not garbage collected;
void <- gc()
stopifnot(nbrOfFreeWorkers() == 1)
```

Now, this is not a big deal, because the worker is indeed available if there is a need for it:

```r
fs <- lapply(1:2, FUN = function(x) { future({ Sys.sleep(x); Sys.getpid() }) })
stopifnot(nbrOfFreeWorkers() == 0)

vs <- value(fs)
stopifnot(nbrOfFreeWorkers() == 2)

message("Worker PIDs: ", paste(vs, collapse = ", "))
#> Worker PIDs: 1101943, 1101942
```

The two different PIDs prove that the two new futures were indeed processed on the two workers.  We also see that there are two free workers available afterwards.

This is the case also for `plan(future.callr::callr, workers = 2)` - also for the current CRAN version of **future.callr**.

**Comment by HenrikBengtsson** [comment]
I think what remains here is to make sure we don't end up in a deadlock from `nbrOfFreeWorkers()` returning 0L, i.e. the idea in https://github.com/futureverse/future.callr/issues/20#issuecomment-1024973101:

> One approach could be to check if nbrOfFreeWorkers() == 0L, then an attempt to garbage collect internal futures should be made before creating the first future.

**Comment by HenrikBengtsson** [comment]
> make sure we don't end up in a deadlock from `nbrOfFreeWorkers()` returning 0L

Even so, as long as one does not do something like `if (nbrOfFreeWorkers() > 0) f <- future(...)`, but instead just go ahead a create the future as `f <- future(...)`, there is already an internal "cleanup" in place, which works by always checking for resolved futures and collecting their results when creating a new future.

That said, for anyone using `if (nbrOfFreeWorkers() > 0) f <- future(...)`, we should probably look into having `nbrOfFreeWorkers()` do the same internal cleanup, before it returns 0.

**Comment by HenrikBengtsson** [comment]
> That said, for anyone using `if (nbrOfFreeWorkers() > 0) f <- future(...)`, we should probably look into having `nbrOfFreeWorkers()` do the same internal cleanup, before it returns 0.

This part is now tracked in https://github.com/futureverse/future/issues/774, which allows me to close this issue.
